File: baekjoon/src/tags/binarySearchh/No10816.java
package tags.binarySearchh;

import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/10816
 */
public class No10816 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static StringTokenizer tokenizer;

    public static int[] cards;
    public static int[] numbers;

    public static void setValues() throws IOException {
        int numberOfInput = Integer.parseInt(reader.readLine());
        tokenizer = new StringTokenizer(reader.readLine());

        cards = new int[numberOfInput];
        for (int i = 0; i < numberOfInput; i++)
            cards[i] = Integer.parseInt(tokenizer.nextToken());
        Arrays.sort(cards);

        numberOfInput = Integer.parseInt(reader.readLine());
        tokenizer = new StringTokenizer(reader.readLine());

        numbers = new int[numberOfInput];
        for (int i = 0; i < numberOfInput; i++)
            numbers[i] = Integer.parseInt(tokenizer.nextToken());
    }

    public static void printCardsCount() throws IOException {
        for (int number : numbers)
            writer.write((getUpperBound(number) - getLowerBound(number)) + " ");
        writer.flush();
    }

    private static int getUpperBound(int number) {
        int low = -1;
        int high = cards.length;

        while (low + 1 < high) {
            int mid = (low + high) / 2;
            if (number >= cards[mid])
                low = mid;
            else
                high = mid;
        }
        return high;
    }

    private static int getLowerBound(int number) {
        int low = -1;
        int high = cards.length;

        while (low + 1 < high) {
            int mid = (low + high) / 2;
            if (number <= cards[mid])
                high = mid;
            else
                low = mid;
        }
        return low + 1;
    }

    public static void main(String[] args) throws IOException {
        setValues();
        printCardsCount();
    }

    /**
     * Map을 이용한 풀이
     *
     * public static void main(String[] args) throws IOException {
     *         Map<Integer, Integer> map = new HashMap<>();
     *         int numberOfInput = Integer.parseInt(reader.readLine());
     *         tokenizer = new StringTokenizer(reader.readLine());
     *         for (int i = 0; i < numberOfInput; i++) {
     *             int number = Integer.parseInt(tokenizer.nextToken());
     *             map.put(number, map.getOrDefault(number, 0) + 1);
     *         }
     *
     *         numberOfInput = Integer.parseInt(reader.readLine());
     *         tokenizer = new StringTokenizer(reader.readLine());
     *         for (int i = 0; i < numberOfInput; i++) {
     *             int number = Integer.parseInt(tokenizer.nextToken());
     *             writer.write(map.getOrDefault(number, 0) + " ");
     *         }
     *         writer.flush();
     *     }
     */
}

File: baekjoon/src/tags/sorting/No10816.kt
package tags.sorting

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/10816
 */
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    reader.readLine().toInt()
    val cards = reader.readLine().split(' ').map { it.toInt() }.sorted()
    reader.readLine()
    val numbers = reader.readLine().split(' ').map { it.toInt() }

    for (number in numbers)
        writer.write("${getUpperBound(cards, number) - getLowerBound(cards, number)} ")
    writer.flush()
}

private fun getUpperBound(cards: List<Int>, number: Int): Int {
    var low = -1
    var high = cards.size
    while (low + 1 < high) {
        val mid = (low + high) / 2
        if (cards[mid] <= number)
            low = mid
        else
            high = mid
    }
    return high
}

private fun getLowerBound(cards: List<Int>, number: Int): Int {
    var low = -1
    var high = cards.size
    while (low + 1 < high) {
        val mid = (low + high) / 2
        if (cards[mid] >= number)
            high = mid
        else
            low = mid
    }
    return high
}

/*
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    reader.readLine().toInt()
    val cards = reader.readLine().split(' ').map { it.toInt() }.sorted()
    reader.readLine()
    val numbers = reader.readLine().split(' ').map { it.toInt() }

    val countOfCards = HashMap<Int, Int>()
    for (card in cards)
        countOfCards[card] = countOfCards.getOrDefault(card, 0) + 1

    for (number in numbers)
        writer.write("${countOfCards.getOrDefault(number, 0)} ")
    writer.flush()
}
*/